File: examples/R/check_response_status.R
#!/usr/bin/env Rscript
suppressMessages(library(scrapystreaming))


status <- parse_input()
create_spider("status", character(0))

requests <- c(200, 201, 400, 404, 500)

for (code in requests) {
    send_request(sprintf("http://httpbin.org/status/%s", code), function(r) {})
}

pending_requests <- length(requests)
results <- list()

while (pending_requests) {
    msg <- parse_input()
    # analyze the responses

    if (msg["type"] == "exception") {
        sent_message <- fromJSON(msg$received_message)
        results[sent_message$url] <- FALSE
    } else {
        results[msg$url] <- TRUE
    }

    pending_requests <- pending_requests - 1
}

f <- file("outputs/response_status.json")
write(toJSON(results), f)
close(f)

File: examples/Nodejs/check_response_status.js
#!/usr/bin/env node

var scrapy = require('scrapystreaming');
var jsonfile = require('jsonfile');

scrapy.createSpider('status', [], function(response) {});

var requests = [200, 201, 400, 404, 500];
var pendingRequests = requests.length;
var result = {};

var canClose = function() {
    if (pendingRequests == 0) {
        jsonfile.writeFile('outputs/check_response.json', result);
        scrapy.closeSpider();
    }
}

var check_response = function(response) {
    result[response.url] = true;
    pendingRequests--;
    canClose();
};

for (req in requests) {
    scrapy.sendRequest('http://httpbin.org/status/' + requests[req], check_response);
}

scrapy.runSpider(function(exception) {
    var msg = JSON.parse(exception.received_message);
    result[msg.url] = false;
    pendingRequests--;
    canClose();
});
